### PR TITLE
[Feature] Rework snapping settings

### PR DESCRIPTION
**Pull request by m-kuhn**
Exposes snapping settings in a new toolbar.
This allows to quickly switch between different snapping modes / settings.
Snapping settings have also been put into a separate object so they can easily be saved / restored during runtime.

This also enables a requirement on [C++ 11 and new-style signal-slot](https://github.com/qgis/qgis3.0_api/issues/35#issuecomment-253805423) this part can be removed if someone asks for it :)
# Toolbar

snapping
# Configuration dialog

screenshot from 2016-10-18 10-06-07
# Snapping settings menu

Has been moved to the Project menu
# TODO
- [x] initial size of layer column + un-collapse tree when first shown
- [x] window has too much padding when served as a dialog
- [x] remove close button
- [ ] move tracing button to snapping toolbar
- [x] remove option to show widget in status bar
- [ ] rework icons
- [ ] button to show legend tree in status bar when in advanced mode

### CODE REVIEW

**Comment by nyalldawson** [comment]
Is this branch really ready for merging? Looks like it could do with a lot of squashing!

**Comment by m-kuhn** [comment]
Yes, squashing is still missing

**Comment by NathanW2** [comment]
Can you post a picture?

**Comment by nirvn** [comment]
@m-kuhn , while your at it and we're allowed to be"disruptive" prior to 3.0, could you move the snapping settings menu shortcut away from settings menu into project menu? It doesn't make sense to have it elsewhere.

**Comment by wonder-sk** [comment]
Good stuff!

Some assorted comments:
- I would suggest to defer switching to new signal/slot connections in one big change for all connections. The old style and new style are not compatible with each other unfortunately, so e.g. disconnecting old style connection with new style will not work...
- would it be possible to move "avoid intersections" and "topo editing" out of snapping config? They are not related to snapping in any way, they are settings respected by some editing map tools. It feels strange to pass those to QgsSnappingUtils. For some unknown reason those extra settings have always been kept alongside snapping config, but I am not sure why...
- I was thinking that snapping "controller" would be a QObject, so it can emit "changed" signal by itself, or process remove layer signals
- SIP: it would be better to put the conversion of QHash to python dict into conversions.sip rather than keeping it embedded in the MethodCode annotation.

**Comment by nyalldawson** [comment]
> I would suggest to defer switching to new signal/slot connections in one big change for all connections. The old style and new style are not compatible with each other unfortunately, so e.g. disconnecting old style connection with new style will not work...

We could use clazy to do this automatically: https://github.com/KDE/clazy/blob/master/checks/level2/README-old-style-connect.md
.... although there is a big warning in that readme.

@m-kuhn 
re the QgsPoint Q_GADGET change - have you tested whether this increases the size of QgsPoint objects? I can't find any info about how much weight this change would make to this class..

**Comment by m-kuhn** [comment]
@nirvn yes, **Move to project menu** sounds good to me

@wonder-sk **New-style signal-slot**. I think it's not so easy.

 * You need to know class-names, so regex will not do it. (No experience with clazy)
 * Some connections will either need to be kept with old style or lambdas (in particular overloaded ones. And I think default arguments sometimes also make it choke.)
 * They can live side-by-side as long as not working on the same connection
 * Therefore I didn't dare to do it in one go
 * At least this all can be sorted out compile time
 * Should have done it separately to keep discussions threaded

**Topo editing / avoid intersections** Any good suggestion where to move them in the UI?

**QObject** The config is a pure data object that can be copied and restored. I think the controller part should go to QgsSnappingUtils which is a QObject.
**QHash/SIP** has been used in other parts too, so some conversion template is in place by default. Not sure why it didn't just work out of the box for this one, maybe because of the nested class.

@nyalldawson **Q_GADGET** [Looks like it adds a staticMetaObject pointer](https://doc-snapshots.qt.io/qt5-dev/qobject.html#Q_GADGET) and I don't think much more. I can check.

**Comment by 3nids** [comment]
@m-kuhn thanks for bringing this to a mergeable state.

* Moving menu to project: :+1: 
* Topo editing / avoid intersection: we can move the settings definition away from QgsProjectSnappingSettings and still keep them in the widget. Or move them to advanced digitizing.
* For Q_OBJECT, @m-kuhn what are the advantages of not being it?

**Comment by m-kuhn** [comment]
> For Q_OBJECT, @m-kuhn what are the advantages of not being it?

It's copyable

**Comment by NathanW2** [comment]
+1 to keeping data just data without being object

**Comment by 3nids** [comment]
OK, we should just make a clear decision on this, as I suppose this will arise soon (it might be worth moving all project properties to some kind of properties registry like this).

**Comment by m-kuhn** [comment]
FYI, in QField I have introduced a second object to manage a `QgsMapSettings` object. https://github.com/opengisch/QField/blob/master/src/mapsettings.h

This way one can have "managed" settings. Often this "manager" object can be integrated into something else (like here the QgsSnappingUtils) and doesn't need to exist explicitely.

**Comment by DelazJ** [comment]
@NathanW2:
 > Can you post a picture?

Yes it would be nice for those interested in the subject but who could not build (you often advise me that, @m-kuhn :) ). Thanks.
Do not also forget tags for the documentation

**Comment by m-kuhn** [comment]
Evaluated the following with and without Q_GADGET:

    QgsPoint p;
    qDebug() << "Sizeof QPoint " << sizeof( p );

    >> Sizeof QPoint 16

**Comment by m-kuhn** [comment]
Toolbar
======
snapping

Configuration dialog
===============
screenshot from 2016-10-18 10-06-07

**Comment by NathanW2** [comment]
Heck yeah. Nice work

**Comment by m-kuhn** [comment]
@wonder-sk , @3nids 

* *Avoid Intersection* settings isolated and created a clean API in QgsProject for it
* *Topological Editing* settings isolated and improve the QgsProject API
* Moved settings to Project menu

File: src/app/qgssnappinglayertreemodel.cpp
Comment (3nids) [inline src/app/qgssnappinglayertreemodel.cpp]: don't you miss to detect changes on `avoidIntersectionsListChanged` ?

**Review by 3nids — APPROVED** [review]

File: src/app/qgssnappinglayertreemodel.cpp
Comment (m-kuhn) [inline src/app/qgssnappinglayertreemodel.cpp]: Yep, fixed.

**Comment by m-kuhn** [comment]
@wonder-sk one less reason to make it a QObject: lambdas in C++ 11 :)

**Comment by DelazJ** [comment]
@m-kuhn Unfortunately, the tags ([Feature] or [needs-docs]) must be in the commit title or text to have it generate an issue in doc repo. Having it in the pull-request seems to have no effect. I also discovered it recently. Thanks for having tried.

That said, could you please create an issue report with the most significant changes in your PR, those valuable for user manual (or cookbook?)? Thanks again.

**Comment by m-kuhn** [comment]
@DelazJ if it's merged via button here (and not via a push), I think it should work, no?

The changes are mainly under the hood (and can be seen in the API docs). Except for those visible in the screenshot on top.

**Comment by DelazJ** [comment]
Ignore (for the moment) the last part of my previous comment. For some reasons, I read in my mailbox that the PR was merged. No need to create an issue in doc if feature not yet exists. Sorry.

> @DelazJ if it's merged via button here (and not via a push), I think it should work, no?

https://github.com/qgis/QGIS/pull/3522 with the tag only in the PR fails last time to create an issue report in doc repo. Reason why I'm not confident in the simple use in the PR. Maybe the webhook (https://github.com/qgis/QGIS-Sysadmin/blob/master/webhooks/github_feature_tracker.cgi#L416) can be improved and accept PR content?

Yes, the valuable changes are those in your first message + the move of snapping settings to project menu (which needs update in doc). All the issue is to have this PR visible in doc repo (we, documenters, become a bit lazy with this nice webhook).

**Comment by m-kuhn** [comment]
What about this one? https://github.com/qgis/QGIS/pull/3597

**Comment by DelazJ** [comment]
the tag is also in the commit

**Comment by m-kuhn** [comment]
That's what I meant, there's a merge-commit produced when clicking the button, depending on the chosen merge-strategy :-)

**Comment by DelazJ** [comment]
However, with PR #3597 committed, we got two reports in the doc repo (which is another issue of cleanup for doc maintainers).

I still do not understand what are the merge strategies :( but as far as you know how things work and can deal with them for the best of us, let's move on. I stop polluting this PR with these considerations.

**Comment by m-kuhn** [comment]
You probably got one from the "real" commit and one from the "merge" commit. But better 2 than 0 ;)

**Comment by DelazJ** [comment]
For sure! But "only 1" is the best.
Looking further the doc issues reported, it appears that double reports happen only if the PR was also a single [Feature] commit. When there's more than one commit (featured or not) in the PR, merging it generates a single report for each concerned commit. If someone want to fix that... :)

**Comment by nyalldawson** [comment]
Just wondering - is it really worth having a status bar mode? The bar is so cluttered already.

**Comment by 3nids** [comment]
I am not really sure about this.
On one hand, it makes sense to have this in the status bar since it's more like behavior control while everything in the toolbar is more action oriented.
On the other hand the toolbar is crowded and having an option to place the widget in status bar or toolbar will decrease coherence between QGIS installations.
Matthias suggested to have control over what is displayed in the status bar.

I would be in favor of keeping it but I have no strong opinion here.

**Comment by nyalldawson** [comment]
It'd be nice to have some way of right-clicking on the status bar widgets and showing/hiding them. Another way we could free up a few pixels would be to swap the "render" checkbox over to a tool button (like the scale lock). Possibly the magnifier and rotation widgets could be hidden behind a button too (ie click the button and the widget pops up in a floating frame about the button).

**Comment by m-kuhn** [comment]
> It'd be nice to have some way of right-clicking on the status bar widgets and showing/hiding them. Another way we could free up a few pixels would be to swap the "render" checkbox over to a tool button (like the scale lock). Possibly the magnifier and rotation widgets could be hidden behind a button too (ie click the button and the widget pops up in a floating frame about the button).

I'd say combine the two: hide action widgets behind buttons AND make it possible to show/hide them.

**Comment by nirvn** [comment]
@m-kuhn , testing this out now, really nice work!

Comments and thoughts:
* the snapping settings window has too much padding when served as a dialog:
untitled
* IMHO, it might be a good time to switch the way it opens to a panel by default
* when you launch a "fresh" QGIS (i.e. no .QGIS3 config file), snapping settings aren't set, and the snapping toolbar is full of icon-less buttons, that needs addressing
* due to the magnet icon being disabled-gray, people will probably be confused for an instant, might be worth adding a bit of color there
* also, on the icon, we'll have to find a way to differentiate that from the tracing icon
* would it make sense for the tracing button to be moved away from the advanced editing into to the snapping toolbar? that would help users understand the inherent link between these two features (ping @wonder-sk :) )
* I absolutely love the new advanced snapping layer list which features symbol icons and groups; that said, when I opened the window, all the groups were collapsed, and the list's Layer column wasn't large enough to read layer/group names, which isn't the best first-experience UX

**Comment by wonder-sk** [comment]
Thanks @m-kuhn and @3nids for this!

Also had similar observations to what @nirvn has posted just above this comment. One more thing is that "Close" button in snapping dialog does not work (I would suggest to get rid of that button completely)

**Comment by nirvn** [comment]
@wonder-sk , good idea re getting rid of the close button (it'd also not be there / needed when in panel mode).

Attaching a better screenshot to demonstrate the dialog padding inconsistency:
untitled

**Comment by 3nids** [comment]
> * IMHO, it might be a good time to switch the way it opens to a panel by default

Is there a consensus on this point?

> * when you launch a "fresh" QGIS (i.e. no QGIS3.conf file), snapping settings aren't set, and the snapping toolbar is full of icon-less buttons, that needs addressing

This is due to remaining old application settings. So, this should not happen with a totally fresh QGIS 3.

> * the above screen grab also demonstrates another issue: due to the magnet icon being disabled-gray, people will probably be confused for an instant, might be worth adding a bit of color there also, on the icon, we'll have to find a way to differentiate that from the tracing icon

I agree these icons might appreciate a bit of attention...

> * would it make sense for the tracing button to be moved away from the advanced editing into to the snapping toolbar? that would help users understand the inherent link between these two features (ping @wonder-sk :) )

might be a good idea. Although, the widget can also be displayed in the statusbar or as a dialog. Tracing action would make sense only in the toolbar I think. There has been some discussion about not showing the widget in the status bar. If we follow this path, let's move tracing icon to the snapping toolbar.

* [ ] window has too much padding when served as a dialog: untitled
* [ ] initial size of layer column + un-collapse tree when first shown
* [ ] remove close button
* [ ] move tracing button to snapping toolbar?
* [ ] rework icons

**Comment by nirvn** [comment]
@3nids , regarding icon-less toolbar buttons, I'm seeing it even on a totally fresh QGIS3, and upon every subsequent QGIS launch.

As for the icons, I can commit myself to rework those if that's OK with you. I'll wait until we have a decision on the statusbar situation.

**Comment by m-kuhn** [comment]
Thanks for the feedback and for the help with the icons.

I'd propose to leave the option for snapping in the status bar, since I think it doesn't hurt (after all it's opt-in), or did I miss something?

**Comment by 3nids** [comment]
@nirvn perfect for the icon, I'll be happy if you do the polish! 

@m-kuhn well, if we allow having it in the status bar, does it make sense to have the tracing icon there?

**Comment by m-kuhn** [comment]
@nirvn @wonder-sk @3nids 
Main UX issues are addressed in latest commit, I think it looks a lot better already, thanks for all the input.

Not sure about the tracing icon (and also no strong opinion about the status bar), does someone want to make a call?

**Comment by NathanW2** [comment]
Nice work. Can u get a screenshot of the status bar?

**Comment by nyalldawson** [comment]
> Main UX issues are addressed in latest commit, I think it looks a lot better already, thanks for all the input.

One last improvement I think could be made - if the snapping mode is set to "advanced configuration" then I think the button which was showing "vertex/segment/..." should change to a settings button to open the advanced snapping settings dialog.

> Not sure about the tracing icon (and also no strong opinion about the status bar), does someone want to make a call?

My opinion: UI controlling options are evil, and we should kill all UI options unless they are REALLY needed. I'd say this isn't really needed.

**Comment by nirvn** [comment]
@nyalldawson , we can't remove the tracing activation icon / option , that is a scenario when it is really needed. 

My vote:
* +1 to move tracing activation icon next to snapping activation icon
* -1 for status bar

**Comment by nyalldawson** [comment]
Oh yeah, I definitely wasn't suggesting removing that! IMO it should be in this toolbar too.

**Comment by nirvn** [comment]
@m-kuhn , nice set of improvements. I still face the issue reported above whereas upon launching QGIS, the snapping toolbar has two icon-less buttons.

**Comment by 3nids** [comment]
ok, +1 to remove it from status bar and move tracing icon.

@nyalldawson I had the same thoughts in mind for advanced configuration. A tool button could allow to open directly the tree under the status bar. I was missing a bit of time for that.

I update the todo list and place it in the main message on top.

**Comment by 3nids** [comment]
@nirvn can you check your default snapping settings in main QGIS options, set them and restart QGIS?

**Comment by m-kuhn** [comment]
* Snapping status bar is gone
* Issue with empty default mode tool button is fixed

**Comment by 3nids** [comment]
@m-kuhn thanks!

**Pull request closed without merging**
